### PR TITLE
chore: enable TypeScript strict mode

### PR DESCRIPTION
**Pull request by addaleax**
Enable TypeScript strict mode for all source files and some tests.

As part of that:

- Update TypeScript itself to the latest version
- Update the eslint plugins for TypeScript to the latest version
- Update the eslint config to account for that
- Update several dependencies to versions that ship TS definitions
- Add devDependencies for several `@types/...` packages
- Update the TypeScript target to ES2018
  - This is much nicer to inspect when looking at the generated code,
    and in particular does not require transpiling async functions

### CODE REVIEW

File: package.json
Comment (addaleax) [inline package.json:7]: I think I accidentally used a newer npm version at some point that did this automatically :woman_shrugging:

File: packages/async-rewriter/tsconfig.json
Comment (addaleax) [inline packages/async-rewriter/tsconfig.json:6]: I’ve skipped the conversion for `async-rewriter` because it was particularly difficult to work with :confused:

File: packages/browser-repl/src/iframe-runtime/iframe-runtime.ts
Comment (addaleax) [inline packages/browser-repl/src/iframe-runtime/iframe-runtime.ts:49]: This might have been a bug? The previous code checks for `this.iframe`, but the presence of the property does not imply that the `onload` event that we wait for below already happened.

File: packages/browser-runtime-core/src/interpreter/interpreter.ts
Comment (addaleax) [inline packages/browser-runtime-core/src/interpreter/interpreter.ts:34]: Still not sure about this one :)

File: packages/browser-runtime-core/src/interpreter/preprocessor/inject-last-expression-callback.ts
Comment (addaleax) [inline packages/browser-runtime-core/src/interpreter/preprocessor/inject-last-expression-callback.ts:5]: `template.statement` is a bit nicer here than `template` because the latter automatically switches between a single `Statement` or an array of them. We’re only dealing with single statements here and can be explicit about that.

File: packages/build/src/barque.spec.ts
Comment (addaleax) [inline packages/build/src/barque.spec.ts:9]: TypeScript has problems understanding that `URL` is a global in Node.js, which is a bit surprising, but I also don’t think there’s any harm in importing it :woman_shrugging:

File: packages/build/src/execute-once.ts
Comment (addaleax) [inline packages/build/src/execute-once.ts:27]: This was dead code, so I removed it :woman_shrugging:

File: packages/cli-repl/src/cli-repl.ts
Comment (addaleax) [inline packages/cli-repl/src/cli-repl.ts:13]: Quoting the README of https://github.com/mcollina/fast-json-parse:

> This assumption holds true up to Node 6, from Node 7 and forward this module is not useful anymore.

File: packages/cli-repl/src/constants.ts
Comment (addaleax) [inline packages/cli-repl/src/constants.ts:41]: Finding typos definitely made this worth it btw :slightly_smiling_face:

File: packages/i18n/src/locales/en_US.ts
Comment (addaleax) [inline packages/i18n/src/locales/en_US.ts:236]: Btw, I was not sure how to actually access the help text from the shell to verify that this was indeed correct, I just inferred that it must be right this way because this is the one where the `attributes` property was on the outer scope.

File: packages/java-shell/src/test/js/run-tests.ts
Comment (addaleax) [inline packages/java-shell/src/test/js/run-tests.ts:8]: Feels like 5 minutes should have been enough, but this timed out once, then I doubled it, then it didn’t time out, soooooo … yeah :woman_shrugging:

File: packages/service-provider-core/src/cursor.ts
Comment (addaleax) [inline packages/service-provider-core/src/cursor.ts:43]: The removed methods here do not actually exist.

File: packages/service-provider-core/src/index.ts
Comment (addaleax) [inline packages/service-provider-core/src/index.ts:18]: Working around https://jira.mongodb.org/browse/NODE-2847

File: packages/service-provider-core/src/result.ts
Comment (addaleax) [inline packages/service-provider-core/src/result.ts:2]: This may be a bit annoying but it’s probably also going to make us better at actually describing the types that we work with :wink:

File: packages/shell-api/src/aggregation-cursor.ts
Comment (addaleax) [inline packages/shell-api/src/aggregation-cursor.ts:73]: `isExhausted` does not exist on the underlying server, so I assumed doing this like we do it for regular shell `Cusor` objects should be fine

File: packages/shell-api/src/aggregation-cursor.ts
Comment (addaleax) [inline packages/shell-api/src/aggregation-cursor.ts:85]: ditto here, there is no `_itcount()` :)

File: packages/shell-api/src/aggregation-cursor.ts
Comment (addaleax) [inline packages/shell-api/src/aggregation-cursor.ts:90]: I’m less sure about these. The fact that these used to return the original cursor and that that was *tested* stumped me a bit, tbh. What would the use of a `.map()` method be that doesn’t actually let you use the mapped results? This was not intended, right? 

Fwiw, this also doesn’t match https://docs.mongodb.com/manual/reference/method/cursor.map/, which claims that this gives an array, but returning a new cursor feels like the most sensible thing to me.

File: packages/shell-api/src/decorators.ts
Comment (addaleax) [inline packages/shell-api/src/decorators.ts:193]: Fun fact: When you transpile from TS to ES6 or above, class properties are non-enumerable, but for ES5 they were, see https://github.com/microsoft/TypeScript/issues/15038.

File: packages/shell-api/src/explainable-cursor.ts
Comment (addaleax) [inline packages/shell-api/src/explainable-cursor.ts:27]: Feel free to tell me if this doesn’t make sense this way :)

File: packages/shell-api/src/integration.spec.ts
Comment (addaleax) [inline packages/shell-api/src/integration.spec.ts:89]: The typings said `res.acknowledged` would be a boolean, the reality said it’s 1/0. I’m happy to make the other change and modify the types instead, if that’s preferable.

**Review by addaleax — COMMENTED** [review]
Leaving a few comments on my own PR because I feel like some things deserve an extra note :)

File: packages/cli-repl/src/cli-repl.ts
Comment (addaleax) [inline packages/cli-repl/src/cli-repl.ts:291]: `Date.now()` returns ms, but `bson.ObjectId()` expects seconds, which is why this started to throw an overflow error in bson 4.2.0 … probably not intended on their side, but a bug on ours, and the `ObjectId` constructor defaults to the current timestamp anyway

File: packages/browser-repl/src/iframe-runtime/iframe-runtime.ts
Comment (mcasimir) [reply packages/browser-repl/src/iframe-runtime/iframe-runtime.ts:49]: I'm only impressed that you managed to spot and fix this weird concurrency issue among a mountain of code 😄 . Yep that could have been a bug, just never happened cause the first evaluation is normally happening quite late from the initialisation. 💯

Comment (addaleax) [reply packages/browser-repl/src/iframe-runtime/iframe-runtime.ts:49]: I really spent far too much time on the Worker thread implementation in Node.js not to be on the lookup for concurrency issues :see_no_evil:

File: packages/browser-runtime-core/src/interpreter/interpreter.ts
Comment (mcasimir) [reply packages/browser-runtime-core/src/interpreter/interpreter.ts:34]: `val` doesn't have to be a `Promise`, the fun part is that i don't know what is making that a `ShellResult` either, essentially that function is going to be injected in the code like this:

```
interpreter.evaluate(`someExpr(); 1 + 1`)
=>

iframe.eval(`someExpr(); ___MONGOSH_LAST_EXPRESSION_CALLBACK (1 + 1)`);
```

I don't think the result there is a `ShellResult`

Comment (addaleax) [reply packages/browser-runtime-core/src/interpreter/interpreter.ts:34]: Ok, then it’s not just me, and the previous code was just wrong…?

File: packages/browser-runtime-core/src/interpreter/interpreter.ts
Comment (mcasimir) [inline packages/browser-runtime-core/src/interpreter/interpreter.ts:29]: I'm not sure about this one either (see the comment about `LAST_EXPRESSION_CALLBACK_FUNCTION_NAME `)

Comment (addaleax) [reply packages/browser-runtime-core/src/interpreter/interpreter.ts:29]: Right … looking into it a bit more, I think this was never supposed to return a `Promise` or a `ShellResult` anyway – it’s all lies… ;)

File: packages/browser-runtime-core/src/interpreter/interpreter.ts
Comment (mcasimir) [reply packages/browser-runtime-core/src/interpreter/interpreter.ts]: I feel sorry about the confusion of this code, i've always thought (i still do :)) that at some point we would have reconciled the whole JS preprocessing (browser friendly adaptations, async-rewrite ...) and JS interpretation layer (IFRAME.eval, vm.runScript, repl eval) and have only one.

I feel the fact that I've used inheritance with the `OpenContextRuntime` deserves an apology on its own. Don't know what got into me, but was completely unnecessary.

File: packages/build/src/build-and-upload.ts
Comment (mcasimir) [inline packages/build/src/build-and-upload.ts:27]: 👍

File: packages/build/src/build-and-upload.ts
Comment (mcasimir) [inline packages/build/src/build-and-upload.ts:37]: why is this necessary shouldn't it  be a string in config?

File: packages/build/src/build-and-upload.ts
Comment (mcasimir) [inline packages/build/src/build-and-upload.ts:53]: same stuff isn't it a string there?

File: packages/cli-repl/src/cli-repl.ts
Comment (mcasimir) [reply packages/cli-repl/src/cli-repl.ts:13]: I mean is also only called once, even if it was faster i'm not sure we need to be that fast :)

File: packages/cli-repl/src/cli-repl.ts
Comment (mcasimir) [inline packages/cli-repl/src/cli-repl.ts:471]: I just noticed the nastiness of this hack for the first time :)

File: packages/i18n/src/locales/en_US.ts
Comment (mcasimir) [reply packages/i18n/src/locales/en_US.ts:236]: Seems right

File: packages/i18n/src/translator.ts
Comment (mcasimir) [inline packages/i18n/src/translator.ts:3]: had no idea this was possible

Comment (addaleax) [reply packages/i18n/src/translator.ts:3]: yeah, it's a relatively recent addition but can be really nice, in particular because the modules are not actually loaded this way

File: packages/shell-api/src/aggregation-cursor.ts
Comment (mcasimir) [reply packages/shell-api/src/aggregation-cursor.ts:90]: I think the original behaviour might have been taken from the driver: https://github.com/mongodb/node-mongodb-native/blob/3.6/lib/cursor.js#L934

Your change makes lot of sense to me, but I think we may need some support from driver in order to make it work or?

File: packages/shell-api/src/cursor.spec.ts
Comment (mcasimir) [inline packages/shell-api/src/cursor.spec.ts:57]: aren't all the tests about `map()`?

File: packages/shell-api/src/database.ts
Comment (mcasimir) [inline packages/shell-api/src/database.ts:787]: I don't know if it applies here, however I think for most of the results coming from the service provider / server a decent compromise between perfectly specifying the shape of the objects and `any` could be something like this:

```
type ServerCommandResult = Record<string, any>;
type IsMasterResult = ServerCommandResult & {
  isMaster: boolean
};

const res1: IsMasterResult = {
  isMaster: true,
  someOtherProp: 123
};

const res2: IsMasterResult = { // Property 'isMaster' is missing ...
  someOtherProp: 123
};
``` 

I'd love to see if the server actually has some IDL files we could steal those interfaces from.

File: packages/shell-api/src/explainable-cursor.ts
Comment (mcasimir) [reply packages/shell-api/src/explainable-cursor.ts:27]: I don't think the current driver explain would work with `map`, but generally with explain you're not interested in the result and you're rather supposed to just call `.explain` at the end. I'd wait on implementing this to see what  driver will do about it, there is a scope (already in progress ?) they are working on.

File: packages/shell-api/src/integration.spec.ts
Comment (mcasimir) [reply packages/shell-api/src/integration.spec.ts:89]: where do we get the types from? The typings from `@types/mongodb` are a bit off, hopefully when the switch of driver to `ts` is done they will be  much more reliable. I'd keep the type from 'reality' anyway :)

File: packages/shell-api/src/shell-api.spec.ts
Comment (mcasimir) [inline packages/shell-api/src/shell-api.spec.ts:160]: is there any other way around `@ts-ignore` here? would be nice to extend the ban to `@ts-ignore` to tests too.

**Review by mcasimir — APPROVED** [review]
... just don't know how you did manage to pull this out in a few days! You even fixed bugs in the meanwhile 😄. I'm super impressed. I've only found few things we should check before merging. 

Thanks for taking care of this, will help a lot in future!

File: packages/build/src/build-and-upload.ts
Comment (addaleax) [reply packages/build/src/build-and-upload.ts:37]: It’s defined as an optional property (which is correct, it’s initialized as `process.env.AWS_KEY` which is a string or `undefined`), so we have to explicitly say it’s a string here … we could also make it always a string and default to an empty one in the build config instead, I guess?

File: packages/shell-api/src/cursor.spec.ts
Comment (addaleax) [reply packages/shell-api/src/cursor.spec.ts:57]: I’m not sure if I’m understanding you correctly… some of the tests here are about `cursor.map()`, yes, but not all of them

File: packages/shell-api/src/shell-api.spec.ts
Comment (addaleax) [reply packages/shell-api/src/shell-api.spec.ts:160]: yeah, this can also use `as any` … kind of comes down to the same thing, I think :)

File: packages/build/src/build-and-upload.ts
Comment (mcasimir) [reply packages/build/src/build-and-upload.ts:37]: I mean you only need that if is unchecked. If we "validate" the value for not being undefined should also change the type in subsequent code.  For example:

```
  config: {
    prop?: string;
  } = {};

  const prop1  = config.prop; // here is string or undefined;

  if (!config.prop) {
    throw new Error('error');
  }

  const prop2 = config.prop; // here is only `string`
```

I think in this case it doesn't work cause is done dynamically:

```
 for (const key of [
    'evgAwsKey', 'evgAwsSecret', 'project', 'revision', 'downloadCenterAwsKey', 'downloadCenterAwsSecret'
  ]) {
    if (typeof (config as any)[key] !== 'string') {
      throw new Error(`Missing build config key: ${key}`);
    }
  }
```

Not sure what's best but leaving it as it is is a good option :).

File: packages/shell-api/src/cursor.spec.ts
Comment (mcasimir) [reply packages/shell-api/src/cursor.spec.ts:57]: i've seen it now, i thought we were doing `describe('method') .. it('does this and that')` everywhere, but it looks like in that file is different.

File: packages/shell-api/src/shell-api.spec.ts
Comment (mcasimir) [reply packages/shell-api/src/shell-api.spec.ts:160]: `@ts-ignore` disables everything, is not going to make much of a difference since have `eslint` catching  issues too, but with `@ts-ignore` for example you could use variables that are not defined anywhere, like `bla + 123` would be compiled just fine.

Comment (mcasimir) [reply packages/shell-api/src/shell-api.spec.ts:160]: but yeah is more of a 'let's shut down the door to bad habits' than something having any practical effect.

Comment (addaleax) [reply packages/shell-api/src/shell-api.spec.ts]: ok, done :)

File: packages/build/src/build-and-upload.ts
Comment (addaleax) [reply packages/build/src/build-and-upload.ts:37]: So you’d suggest testing all of the keys manually with individual `if`s? I think I would prefer not to do that … :smile:

File: packages/shell-api/src/aggregation-cursor.ts
Comment (addaleax) [reply packages/shell-api/src/aggregation-cursor.ts]: Ok, this got cleared up, `.map()` actually does something that does not match my intuition for that method name at all :)

**Comment by addaleax** [comment]
@mcasimir Do we want to merge this at this point?

File: packages/shell-api/src/shell-bson.ts
Comment (gribnoysup) [inline packages/shell-api/src/shell-bson.ts:185]: Maybe we can reduce the amount of `any`s here also with the help of `keyof`:

```ts
// (GitHub doesn't allow me to do a suggestion for all touched lines so just posting a snippet, sorry!)

const keys: (keyof typeof bsonPkg)[] = [
  "ObjectId",
  "Code",
  "DBRef",
  "MaxKey",
  "MinKey",
  "Timestamp",
  "Symbol",
  "Map",
];

keys.forEach((className) => {
  bsonPkg[className].help = (): Help => helps[className];
  Object.setPrototypeOf(bsonPkg[className].help, helps[className]);
});

```

What do you think?

**Review by gribnoysup — APPROVED** [review]
Wow, amazing and thorough work and was very insightful for me to see literally almost every code file in the repository, thank you! 😊 I left a few tiny questions and code suggestions along the way

File: packages/shell-api/src/shell-bson.ts
Comment (addaleax) [reply packages/shell-api/src/shell-bson.ts:185]: And adopted this as well :slightly_smiling_face:

File: packages/browser-runtime-core/src/interpreter/preprocessor/collect-top-level-lexical-context.ts
Comment (addaleax) [reply packages/browser-runtime-core/src/interpreter/preprocessor/collect-top-level-lexical-context.ts:26]: No, you’re right, these can be dropped here :) Thanks, done!

File: packages/build/src/s3.ts
Comment (addaleax) [reply packages/build/src/s3.ts:16]: This seems just fine to me, yes :) Done!

File: packages/cli-repl/src/arg-mapper.ts
Comment (addaleax) [reply packages/cli-repl/src/arg-mapper.ts]: Sure, this works for me just fine, done! :)

File: packages/build/src/build-and-upload.ts
Comment (rose-m) [reply packages/build/src/build-and-upload.ts:37]: Another solution might be to use the `!` operator, so for example we could use `config.evgAwsKey!` instead of `config.evgAwsKey as string` which makes it easier to read IMHO. What do you think?

File: packages/build/src/download-center.spec.ts
Comment (rose-m) [inline packages/build/src/download-center.spec.ts:3]: Interestingly - removing it locally still worked for me... 🤔

**Comment by gribnoysup** [comment]
Oh no, I think my `s3.ts` change suggestion broke the `s3.spec.ts` a little, sorry! 🙈

**Comment by addaleax** [comment]
@gribnoysup Already on it :) I’ll merge this once CI is green :+1:

**Review by rose-m — COMMENTED** [review]
What a huge amount of work - but definitely worth it, nice!! 🥳 

I've written some minor comments and by far don't yet understand anything too much so I'll not "Approve" but just "Comment".

File: packages/build/src/download-center.spec.ts
Comment (addaleax) [reply packages/build/src/download-center.spec.ts:3]: idk, in the end I’m perfectly fine with explicitly importing it either way

File: packages/build/src/build-and-upload.ts
Comment (addaleax) [reply packages/build/src/build-and-upload.ts:37]: I avoided that because it leads to a warning by eslint :woman_shrugging:

**Pull request merged**
